### PR TITLE
[seraphis] seraphis_crypto: add seraphis transcript utility

### PR DESCRIPTION
**Pull request by UkoeHB**
## Summary

This is a PR in my 'upstreaming seraphis_lib project', it is not used anywhere yet.

- Adds `src/seraphis_crypto` directory. Includes a dummy cpp file so cmake won't complain (will remove it once I have a cpp file to add).
- Adds `sp_transcript.h` to `seraphis_crypto`.

## Explanation

In the current codebase, transcripting of crypto protocols is done manually which is verbose and inflexible. I wrote a transcripting utility that is quite ergonomic and has served me well in the seraphis library.

There are two modes
- `SpFSTranscript`: Fiat-Shamir transcript, includes labels, type flags, and type lengths. Suited for variable-length transcripts. Should be used for Fiat-Shamir challenges.
- `SpKDFTranscript`: KDF transcript, does not include labels, type flags, or type lengths. Suited for short fixed-length transcripts (e.g. key derivations). In my seraphis library all uses of KDF transcripts are less than 128 bytes (or 256 bytes at least, can't remember).

Example use:
```cpp
// my_proof.h
struct MyProof
{
    rct::key A;
};
inline const boost::string_ref container_name(const MyProof&) { return "MyProof"; }
void append_to_transcript(const MyProof &container, SpTranscriptBuilder &transcript_inout);

// my_proof.cpp
void append_to_transcript(const MyProof &container, SpTranscriptBuilder &transcript_inout)
{
    transcript_inout.append("A", container.A);
}

// build a transcript and consume it
rct::key make_challenge(const proof &P)
{
    // hash data
    SpFSTranscript transcript{config::HASH_KEY_MY_CHALLENGE, sizeof(rct::key)};
    transcript.append("proof", proof);

    // challenge
    rct::key challenge;
    sp_hash_to_scalar(transcript.data(), transcript.size(), challenge.bytes);

    return challenge;
}
```

### CODE REVIEW

**Review by vtnerd — CHANGES_REQUESTED** [review]
Looks good, but there is one thing that definitely needs to be changed (the signed integer code).

**Review by vtnerd — CHANGES_REQUESTED** [review]
Oops, I thought this wasn't  submitted.

File: src/seraphis_crypto/sp_transcript.h [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+        {
+            // positive integer: byte{0} || varint(uint(int_variable))
+            this->append_uint(0);
+            this->append_uint(static_cast<std::uint64_t>(signed_integer));
+        }
+        else
+        {
+            // negative integer: byte{1} || varint(uint(abs(int_variable)))
+            this->append_uint(1);
+            this->append_uint(static_cast<std::uint64_t>(-signed_integer));
```
[the remark this answers is not in the document]
Comment (vtnerd) [reply src/seraphis_crypto/sp_transcript.h]: Oh I didn't think of that. You'd have a positive and negative zero.

**Comment by selsta** [comment]
@vtnerd could you re-review and potentially approve?

File: src/seraphis_crypto/sp_transcript.h [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Transcript class for assembling data that needs to be hashed.
+
+#pragma once
+
+//local headers
+#include "crypto/crypto.h"
+#include "crypto/x25519.h"
```
Comment (vtnerd) [inline src/seraphis_crypto/sp_transcript.h:35]: I can't find this header, is this new? The checks still pass, so I must be missing something.

**Review by vtnerd — APPROVED** [review]

**Review by selsta — APPROVED** [review]

**Pull request merged**